testsuite/tests/typing-modes/yielding.ml

```diff
@@ -78,3 +72,197 @@ Line 2, characters 45-46:
+  fun f -> f ((:=) storage)
+
+[%%expect{|
+type 'a t1 = Mk1 of global_ 'a
+type 'a t2 = Mk2 of global_ 'a @@ yielding
+type 'a t3 = Mk3 of 'a @@ unyielding
+type 'a t4 = Mk4 of 'a
+val with_global_effect : ((string -> unit) @ yielding -> 'a) -> 'a = <fun>
+|}]
+
+(* [global] modality implies [unyielding]. *)
+let _ = with_global_effect (fun k -> let _ = Mk1 k in ())
+
+[%%expect{|
+Line 1, characters 49-50:
+1 | let _ = with_global_effect (fun k -> let _ = Mk1 k in ())
+                                                     ^
+Error: This value is "yielding" but expected to be "unyielding".
+|}]
+
+(* [global yielding] works: *)
+let _ = with_global_effect (fun k -> let _ = Mk2 k in ())
+
+[%%expect{|
+- : unit = ()
+|}]
+
+(* [unyielding] and [yielding] modalities: *)
+let _ = with_global_effect (fun k -> let _ = Mk3 k in ())
+
+[%%expect{|
+Line 1, characters 49-50:
+1 | let _ = with_global_effect (fun k -> let _ = Mk3 k in ())
+                                                     ^
+Error: This value is "yielding" but expected to be "unyielding".
+|}]
+
+let _ = with_global_effect (fun k -> let _ = Mk4 k in ())
+
+[%%expect{|
+- : unit = ()
+|}]
+
+(* Externals and [yielding]: *)
+
+external ok_yielding : 'a @ local -> unit = "%ignore"
+
+let _ = ok_yielding 4
+
+let _ = ok_yielding (stack_ (Some "local string"))
+
+let _ = with_global_effect (fun k -> ok_yielding k)
+
+[%%expect{|
+external ok_yielding : local_ 'a -> unit = "%ignore"
+- : unit = ()
+- : unit = ()
+- : unit = ()
+|}]
+
+external requires_unyielding : 'a @ local unyielding -> unit = "%ignore"
+
+let _ = requires_unyielding 4
+
+let _ = requires_unyielding (stack_ (Some "local string"))
+
+let _ = with_global_effect (fun k -> requires_unyielding k)
+
+[%%expect{|
+external requires_unyielding : local_ 'a @ unyielding -> unit = "%ignore"
+- : unit = ()
+- : unit = ()
+Line 7, characters 57-58:
+7 | let _ = with_global_effect (fun k -> requires_unyielding k)
+                                                             ^
+Error: This value is "yielding" but expected to be "unyielding".
+|}]
+
+external returns_unyielding : 'a -> 'a @ local unyielding = "%identity"
+
+let _ = requires_unyielding (returns_unyielding "some string")
+
+[%%expect{|
+external returns_unyielding : 'a -> local_ 'a @ unyielding = "%identity"
+- : unit = ()
+|}]
+
+(* [@local_opt] and [yielding]: *)
+
+external id : ('a[@local_opt]) -> ('a[@local_opt]) = "%identity"
+
+let f1 x = id x
+let f2 (x @ local) = exclave_ id x
+let f3 (x @ yielding) = id x
+let f4 (x @ local unyielding) = exclave_ id x
+
+[%%expect{|
+external id : ('a [@local_opt]) -> ('a [@local_opt]) = "%identity"
+val f1 : 'a -> 'a = <fun>
+val f2 : local_ 'a -> local_ 'a = <fun>
+val f3 : 'a @ yielding -> 'a @ yielding = <fun>
+val f4 : local_ 'a @ unyielding -> local_ 'a = <fun>
+|}]
+
+(* Test [instance_prim] + mixed mode annots. *)
+external requires_unyielding : 'a @ local unyielding -> (unit [@local_opt]) = "%ignore"
```

Comment:
Another test case to add is: `'a @ local yielding [@local_opt]` - both `local` and `yielding` are overwritten by the `local_opt` without warning. We don't need to fix it since it was already broken, but should add a test case showing that.

Reply:
Done.